### ISSUE TITLE
Segmentation fault with single signout  in Version 0.12.0

### ISSUE THREAD

**Issue by rsabir** [UNINETT/mod_auth_mellon#78]
I'm getting an segmentation fault when I try to use assertion in Single Singout. What I use is **WSO2 IS  5.1 as IDP, Mod_auth_mellon** and **Centos 6** as OS. Furthermore when I disable verification of signature, the segmentation fault disappears but the mod_auth_mellon doesn't destroy the remote session (in IDP). So what happens at this case is that the session in Apache is destroyed and mod_auth_mellon will ask me to get an SAMLResponse from idp but I won't be promped in IDP for my credentials.
Are these two behaviours bugs in mod_auth_mellon?

**Comment by olavmrk** [comment]
It is hard to tell what actually happens here. The last issue (you not being asked for credentials after logging out) sounds like a configuration error or a bug in the IdP. The first issue (a segmentation fault) is either a bug in mod_auth_mellon or in the Lasso library. Without more information it is hard to tell.

Information that would be useful:
1. The request that triggers the segmentation fault.
2. A stack trace for the segmentation fault.
3. Information about what version of the Lasso library you are using.

**Comment by rsabir** [comment]
Here's the informations asked : 

1.  SAML Request : 

```
<samlp:LogoutRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="_A651BC84E8BFD52623DEA213F871A26E" Version="2.0" IssueInstant="2016-05-17T09:07:56Z" Destination="https://wso2.capgemini.com:9443/samlsso">
<saml:Issuer>mellon</saml:Issuer><saml:NameID Format="urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress">rsabir@carbon.super</saml:NameID><samlp:SessionIndex>ce7b1085-9f75-4cd0-8e9b-6b4090383812</samlp:SessionIndex>
</samlp:LogoutRequest>
```
 SAMLResponse : 

```
<?xml version="1.0" encoding="UTF-8"?>
<saml2p:LogoutResponse Destination="https://mellon.capgemini-sp.com/wso2/auth/mellon/logout" ID="oennmhoklpmpooocddmgpdhpijkdpaodhindpepo" InResponseTo="_A651BC84E8BFD52623DEA213F871A26E" IssueInstant="2016-05-17T09:07:56.914Z" Version="2.0"
    xmlns:saml2p="urn:oasis:names:tc:SAML:2.0:protocol">
    <saml2:Issuer Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity"
        xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">wso2
    </saml2:Issuer>
    <ds:Signature
        xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
        <ds:SignedInfo>
            <ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
            <ds:SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/>
            <ds:Reference URI="#oennmhoklpmpooocddmgpdhpijkdpaodhindpepo">
                <ds:Transforms>
                    <ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
                    <ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
                </ds:Transforms>
                <ds:DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
                <ds:DigestValue>IYKUxGuIc8vGcKYM1rcaeen7GC4=</ds:DigestValue>
            </ds:Reference>
        </ds:SignedInfo>
        <ds:SignatureValue>DZfilyfKOeDg5HQLcM6P0uHAwaTxl37xVqdKlPsFeOZRLrtQddzQacgdWEnNA5i4WD8PZA68sQOF7siUkxMvRl/d4KddLP/BDR+i0G0D8zJoMeeTY6T22YN7AXAfQgQ8d73PeNuDkK75T8AFjCCtKjS/wAQvedx08RXnGIN4NN4=</ds:SignatureValue>
        <ds:KeyInfo>
            <ds:X509Data>
                <ds:X509Certificate>MIICNTCCAZ6gAwIBAgIES343gjANBgkqhkiG9w0BAQUFADBVMQswCQYDVQQGEwJVUzELMAkGA1UECAwCQ0ExFjAUBgNVBAcMDU1vdW50YWluIFZpZXcxDTALBgNVBAoMBFdTTzIxEjAQBgNVBAMMCWxvY2FsaG9zdDAeFw0xMDAyMTkwNzAyMjZaFw0zNTAyMTMwNzAyMjZaMFUxCzAJBgNVBAYTAlVTMQswCQYDVQQIDAJDQTEWMBQGA1UEBwwNTW91bnRhaW4gVmlldzENMAsGA1UECgwEV1NPMjESMBAGA1UEAwwJbG9jYWxob3N0MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCUp/oV1vWc8/TkQSiAvTousMzOM4asB2iltr2QKozni5aVFu818MpOLZIr8LMnTzWllJvvaA5RAAdpbECb+48FjbBe0hseUdN5HpwvnH/DW8ZccGvk53I6Orq7hLCv1ZHtuOCokghz/ATrhyPq+QktMfXnRS4HrKGJTzxaCcU7OQIDAQABoxIwEDAOBgNVHQ8BAf8EBAMCBPAwDQYJKoZIhvcNAQEFBQADgYEAW5wPR7cr1LAdq+IrR44iQlRG5ITCZXY9hI0PygLP2rHANh+PYfTmxbuOnykNGyhM6FjFLbW2uZHQTY1jMrPprjOrmyK5sjJRO4d1DeGHT/YnIjs9JogRKv4XHECwLtIVdAbIdWHEtVZJyMSktcyysFcvuhPQK8Qc/E/Wq8uHSCo=</ds:X509Certificate>
            </ds:X509Data>
        </ds:KeyInfo>
    </ds:Signature>
    <saml2p:Status>
        <saml2p:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
    </saml2p:Status>
</saml2p:LogoutResponse>
```

2. In the error_log, I have only this : 
```
(process:11137): Lasso-CRITICAL **: 2016-05-17 11:07:57 (tools.c/:1165) Failed to inflate
[Tue May 17 11:07:57 2016] [notice] child pid 11137 exit signal Segmentation fault (11)
```
3. I'm using liblasso.so.3 so I assume it's the version 3. I found also in my system (with find command) liblasso.so.3.10.0 so I assume that I use the version 3.10.0

**Comment by olavmrk** [comment]
Based on the contents of the log message from Lasso ("Failed to inflate"), and that the contents of the logout response contains several things that shouldn't be there (signature, certificate, etc), it looks like this logout response is meant for the HTTP-POST binding.

However, mod_auth_mellon does not support HTTP-POST for logout messages. It won't even try to decode them, which indicates that this logout message was actually sent in a query parameter (as if it was using the HTTP-Redirect binding).

In other words, something strange is going on at the IdP. Nevertheless, mod_auth_mellon shouldn't crash when this happens. I was able to reproduce the problem, and have committed a fix for the crash in 6bdda91. mod_auth_mellon now correctly returns a "400 Bad Request" error in this case.

To actually solve your problem (logout not working), you will have to find out why the IdP is sending an incorrectly encoded logout message to the SP.

**Comment by rsabir** [comment]
Ok, so I need to configure the IDP so it uses HTTP-REDIRECT in logout?

**Comment by olavmrk** [comment]
Yes, though as I mentioned it doesn't appear to use HTTP-POST now -- as far as I can tell it uses HTTP-POST encoding with HTTP-Redirect transmission.

HTTP-POST is supposed to be sent using the POST request type, with the message in the POST body, while HTTP-Redirect is supposed to be DEFLATE-compressed and sent as a query string parameter. You appear to receive the message as a query parameter, but without valid DEFLATE-compression.
